File: inbadders.lua
-- Main Love2D script
function love.load()
    -- Constants
    CANVASWIDTH = 800
    CANVASHEIGHT = 600

    -- Key constants
    KEY_LEFT = "left"
    KEY_RIGHT = "right"
    KEY_SPACE = "space"

    -- Initialize game variables
    SPACESHIP = { x = CANVASWIDTH / 2 - 20, y = CANVASHEIGHT - 50 }
    INVADER = { x = 50, y = 50, speed = 2, direction = 1 }
    BULLETS = {}
    for i = 1, 10 do
        BULLETS[i] = { x = 0, y = 0, active = false }
    end
    GAMEOVER = false
    SCORE = 0

    -- Load sprites
    SPACESHIP.sprite = love.graphics.newImage("spaceship.png")
    INVADER.sprite = love.graphics.newImage("invader.png")
    
    -- Create sprites from pixmap definitions
    createSprites()
end

function createSprites()
    -- Define colors according to Pico8 palette
    local colors = {
        [0] = { 0, 0, 0 },
        [1] = { 29, 43, 83 },
        [2] = { 126, 37, 83 },
        [3] = { 0, 135, 81 },
        [4] = { 171, 82, 54 },
        [5] = { 95, 87, 79 },
        [6] = { 194, 195, 199 },
        [7] = { 255, 241, 232 },
        [8] = { 255, 0, 77 },
        [9] = { 255, 163, 0 },
        [10] = { 255, 236, 39 },
        [11] = { 0, 228, 54 },
        [12] = { 41, 173, 255 },
        [13] = { 131, 118, 156 },
        [14] = { 255, 119, 168 },
        [15] = { 255, 204, 170 }
    }

    -- Create spaceship sprite
    SPACESHIP.pixels = {
        {0,0,12, 7,  7, 12, 0, 0},
        {0,7,  7,  7,  7,  7, 7, 0},
        {12,7,  7,  7,  7,  7, 7, 12},
        {12,7,  0,  7,  7,  0, 7, 12},
        {12,0,  0,  7,  7,  0, 0, 12}
    }
    SPACESHIP.image = createSpriteImage(SPACESHIP.pixels, colors)

    -- Create invader sprite
    INVADER.pixels = {
        {0, 2, 2, 2, 2, 2, 2, 0},
        {2, 2, 0, 0, 0, 0, 2, 2},
        {2, 2, 11, 11, 11, 11, 2, 2},
        {2, 0, 2, 2, 2, 2, 0, 2},
        {2, 0, 2, 0, 0, 2, 0, 2}
    }
    INVADER.image = createSpriteImage(INVADER.pixels, colors)
end

function createSpriteImage(pixels, colors)
    local width, height = #pixels[1], #pixels
    local imageData = love.image.newImageData(width, height)

    for y = 1, height do
        for x = 1, width do
            local colorIndex = pixels[y][x]
            if colorIndex > 0 then
                local r, g, b = unpack(colors[colorIndex])
                imageData:setPixel(x - 1, y - 1, r / 255, g / 255, b / 255, 1)
            end
        end
    end

    return love.graphics.newImage(imageData)
end

function love.update(dt)
    if not GAMEOVER then
        moveSpaceship()
        moveInvader()
        moveBullets()
        checkGameOver()
    end
end

function love.draw()
    if GAMEOVER then
        drawGameOver()
    else
        love.graphics.clear(0, 0, 0)
        love.graphics.setColor(1, 1, 1)
        love.graphics.draw(SPACESHIP.image, SPACESHIP.x, SPACESHIP.y, 0, 8, 8)
        love.graphics.draw(INVADER.image, INVADER.x, INVADER.y, 0, 8, 8)
        drawBullets()
        drawScore()
    end
end

function drawScore()
    love.graphics.setColor(1, 1, 1)
    love.graphics.print("SCORE: " .. SCORE, 10, 10)
end

function drawGameOver()
    love.graphics.setColor(1, 1, 1)
    love.graphics.printf("GAME OVER", 0, CANVASHEIGHT / 2 - 30, CANVASWIDTH, "center")
    love.graphics.printf("SCORE: " .. SCORE, 0, CANVASHEIGHT / 2 + 10, CANVASWIDTH, "center")
end

function moveSpaceship()
    if love.keyboard.isDown(KEY_LEFT) then
        SPACESHIP.x = math.max(SPACESHIP.x - 5, 0)
    end
    if love.keyboard.isDown(KEY_RIGHT) then
        SPACESHIP.x = math.min(SPACESHIP.x + 5, CANVASWIDTH - 40)
    end
    if love.keyboard.isDown(KEY_SPACE) then
        shootBullet()
    end
end

function shootBullet()
    for i = 1, #BULLETS do
        if not BULLETS[i].active then
            BULLETS[i].x = SPACESHIP.x + 20
            BULLETS[i].y = SPACESHIP.y - 10
            BULLETS[i].active = true
            break
        end
    end
end

function moveBullets()
    for i = 1, #BULLETS do
        if BULLETS[i].active then
            BULLETS[i].y = BULLETS[i].y - 5
            if BULLETS[i].y < 0 then
                BULLETS[i].active = false
            elseif BULLETS[i].y < INVADER.y + 5 and BULLETS[i].x > INVADER.x and BULLETS[i].x < INVADER.x + 40 then
                BULLETS[i].active = false
                SCORE = SCORE + 100
                INVADER.y = 600 -- Move invader off-screen
            end
        end
    end
end

function drawBullets()
    for i = 1, #BULLETS do
        if BULLETS[i].active then
            love.graphics.setColor(1, 1, 1)
            love.graphics.rectangle("fill", BULLETS[i].x, BULLETS[i].y, 2, 10) -- Draw bullet
        end
    end
end

function moveInvader()
    INVADER.x = INVADER.x + INVADER.speed * INVADER.direction
    if INVADER.x < 0 or INVADER.x > CANVASWIDTH - 40 then
        INVADER.direction = -INVADER.direction
        INVADER.y = INVADER.y + 10 -- Move invader down
    end
end

function checkGameOver()
    if INVADER.y > CANVASHEIGHT then
        GAMEOVER = true
    end
end

function love.keypressed(key)
    if GAMEOVER and key == "return" then
        -- Restart the game
        love.load()
    end
end

File: inbadders.py
import pygame
import sys
import random

# Constants
CANVASWIDTH = 800
CANVASHEIGHT = 600
KEY_LEFT = 37
KEY_RIGHT = 39
KEY_SPACE = 32

# Pico8 Palette
PALETTE = [
    (0, 0, 0),      # 0: black
    (29, 43, 83),   # 1: dark-blue
    (126, 37, 83),  # 2: dark-purple
    (0, 135, 81),   # 3: dark-green
    (171, 82, 54),  # 4: brown
    (95, 87, 79),   # 5: dark-grey
    (194, 195, 199),# 6: light-grey
    (255, 241, 232),# 7: white
    (255, 0, 77),   # 8: red
    (255, 163, 0),  # 9: orange
    (255, 236, 39), # 10: yellow
    (0, 228, 54),   # 11: green
    (41, 173, 255), # 12: blue
    (131, 118, 156),# 13: lavender
    (255, 119, 168),# 14: pink
    (255, 204, 170) # 15: light-peach
]

# Sprite definitions
SPACESHIP_PIXMAP = [
    [0, 0, 12, 7, 7, 12, 0, 0],
    [0, 7, 7, 7, 7, 7, 7, 0],
    [12, 7, 7, 7, 7, 7, 7, 12],
    [12, 7, 0, 7, 7, 0, 7, 12],
    [12, 0, 0, 7, 7, 0, 0, 12]
]

INVADER_PIXMAP = [
    [0, 2, 2, 2, 2, 2, 2, 0],
    [2, 2, 0, 0, 0, 0, 2, 2],
    [2, 2, 11, 11, 11, 11, 2, 2],
    [2, 0, 2, 2, 2, 2, 0, 2],
    [2, 0, 2, 0, 0, 2, 0, 2]
]

# Bullet definition
class Bullet:
    def __init__(self):
        self.x = 0
        self.y = 0
        self.active = False

# Game class
class Game:
    def __init__(self):
        self.spaceship_x = CANVASWIDTH / 2 - 20
        self.spaceship_y = CANVASHEIGHT - 50
        self.invader_x = 50
        self.invader_y = 50
        self.invader_speed = 2
        self.invader_direction = 1  # 1 for right, -1 for left
        self.game_over = False
        self.score = 0
        self.bullets = [Bullet() for _ in range(10)]
    
    def draw_sprite(self, surface, pixmap, x, y):
        for row_idx, row in enumerate(pixmap):
            for col_idx, color_index in enumerate(row):
                if color_index != 0:  # Assuming 0 is transparent
                    pygame.draw.rect(surface, PALETTE[color_index], 
                                     (x + col_idx * 8, y + row_idx * 8, 8, 8))

    def draw_game_over(self, surface):
        font = pygame.font.SysFont("Arial", 30)
        text = font.render("GAME OVER", True, PALETTE[7])
        surface.blit(text, (CANVASWIDTH / 2 - text.get_width() / 2, CANVASHEIGHT / 2))
        
        score_text = font.render(f"SCORE: {self.score}", True, PALETTE[7])
        surface.blit(score_text, (CANVASWIDTH / 2 - score_text.get_width() / 2, CANVASHEIGHT / 2 + 40))

    def draw_score(self, surface):
        font = pygame.font.SysFont("Arial", 20)
        score_text = font.render(f"SCORE: {self.score}", True, PALETTE[7])
        surface.blit(score_text, (10, 10))

    def move_bullets(self):
        for bullet in self.bullets:
            if bullet.active:
                bullet.y -= 5  # Move bullet up
                if bullet.y < 0:
                    bullet.active = False
                if bullet.active and bullet.y < self.invader_y + 5 and \
                   self.invader_x < bullet.x < self.invader_x + 40:
                    bullet.active = False  # Deactivate bullet
                    self.score += 100      # Increase score
                    self.invader_y = 600   # Move invader off-screen

    def move_spaceship(self):
        keys = pygame.key.get_pressed()
        if keys[pygame.K_LEFT]:
            self.spaceship_x -= 5
        if keys[pygame.K_RIGHT]:
            self.spaceship_x += 5
        if keys[pygame.K_SPACE]:
            self.shoot_bullet()

    def shoot_bullet(self):
        for bullet in self.bullets:
            if not bullet.active:
                bullet.x = self.spaceship_x + 20
                bullet.y = self.spaceship_y - 10
                bullet.active = True
                break

    def move_invader(self):
        self.invader_x += self.invader_speed * self.invader_direction
        if self.invader_x < 0 or self.invader_x > CANVASWIDTH - 40:
            self.invader_direction = -self.invader_direction
            self.invader_y += 10  # Move invader down

    def check_game_over(self):
        if self.invader_y > CANVASHEIGHT:
            self.game_over = True

    def run(self):
        pygame.init()
        screen = pygame.display.set_mode((CANVASWIDTH, CANVASHEIGHT))
        clock = pygame.time.Clock()

        while True:
            for event in pygame.event.get():
                if event.type == pygame.QUIT:
                    pygame.quit()
                    sys.exit()

            if not self.game_over:
                screen.fill(PALETTE[0])  # Clear screen with black
                self.draw_score(screen)
                self.draw_sprite(screen, SPACESHIP_PIXMAP, self.spaceship_x, self.spaceship_y)
                self.draw_sprite(screen, INVADER_PIXMAP, self.invader_x, self.invader_y)
                
                for bullet in self.bullets:
                    if bullet.active:
                        self.draw_sprite(screen, [[11]], bullet.x, bullet.y)

                self.move_spaceship()
                self.move_bullets()
                self.move_invader()
                self.check_game_over()
            else:
                self.draw_game_over(screen)

            pygame.display.flip()
            clock.tick(60)  # Frame rate

if __name__ == "__main__":
    Game().run()

